### ISSUE TITLE
User-defined default selection criteria

### ISSUE THREAD

**Issue by jtcohen6** [dbt-labs/dbt#3448]
_Credit where due: This was @aescay's idea many months ago!_

### Describe the feature

We've increasingly heard a desire from community members to limit `dbt run` in development to only a subset of their project, or to always exclude a set of infrequently run models. Trying to solve for this with `enabled` gets tricky quickly, because disabled resources cannot participate in the DAG and will raise dependency errors accordingly.

Instead, what if there were a way to redefine the default selection criteria? I think the right construct for this is a [yaml selector](https://docs.getdbt.com/reference/node-selection/yaml-selectors) (as in `selectors.yml`). We could support a new selector property, `default: true|false`, or check for a selector named `default`. The former is almost certaintly better, we'd just need to raise an error if multiple selectors have `default: true`.

If specified, that selector should override the default includes + excludes defined here:

https://github.com/fishtown-analytics/dbt/blob/abe8e839458d0288f76cb6950e79c77e3b1627cc/core/dbt/graph/cli.py#L22-L23

Then, `dbt run` would have the same effect as `dbt run --selector my_default_selector`.

### Questions

1. How to combine the user-defined default with additional selection criteria passed via CLI syntax? Should the two be combined, or should the CLI criteria entirely override the default? I'm leaning toward total override for two reasons:
    - It's what happens today: `dbt ls` really means ` dbt ls --select fqn:* source:* exposure:*`, as soon as the user says `-s something_else`, the command becomes `dbt ls -s something_else`
    - Override feels like the only way to make sense of `dbt ls --selector not_the_default`
2. Can we support different default selectors in different environments? I think it makes a lot of sense to perma-exclude certain models in development, but still select them in production. As it turns out, `selectors.yml` already supports Jinja today, so this could be as simple as:
```yml
selectors:
  - name: prod
    description: Select everything in prod
    default: "{{ target.name == 'prod' | as_bool }}"
    definition: 'fqn:* source:* exposure:*'
  - name: dev
    description: Avoid unpleasant surprises in dev
    default: "{{ target.name == 'dev' | as_bool }}"
    definition:
      union:
        - 'fqn:* source:* exposure:*'
        - exclude:
            - unpleasant
            - surprises
```

**Comment by TeddyCr** [comment]
Hey @jtcohen6 I would love to tackle this one. 🙂

**Comment by jtcohen6** [comment]
@TeddyCr I was thinking that the approach here would be to check `selectors` and override the [default includes + excludes](https://github.com/dbt-labs/dbt/blob/abe8e839458d0288f76cb6950e79c77e3b1627cc/core/dbt/graph/cli.py#L22-L23), which are used as inputs to [`parse_difference`](https://github.com/dbt-labs/dbt/blob/abe8e839458d0288f76cb6950e79c77e3b1627cc/core/dbt/graph/cli.py#L69-L70) lower down.

The way this works in practice, is that a task calls `get_selection_spec`. First it checks to see if a `--selector` was passed, otherwise, it passes `--models`/`--select` and `--exclude` into `parse_difference`:

https://github.com/dbt-labs/dbt/blob/45fe76eef4b4b82ae1442f00310e0c6a121774f2/core/dbt/task/list.py#L171-L176

https://github.com/dbt-labs/dbt/blob/45fe76eef4b4b82ae1442f00310e0c6a121774f2/core/dbt/task/compile.py#L40-L45

https://github.com/dbt-labs/dbt/blob/45fe76eef4b4b82ae1442f00310e0c6a121774f2/core/dbt/task/freshness.py#L139-L150

So now I'm thinking a better approach here may be to add a check there: If none of `--selector_name`, `--models`/`--select`, and `--exclude` is set, then `self.config.get_selector('default')`:
```python
 def get_selection_spec(self) -> SelectionSpec: 
     if self.args.selector_name: 
         spec = self.config.get_selector(self.args.selector_name)
     elif not (self.args.models or self.args.exclude):
         # grab by new 'default' property rather than name
         # need some way to check that a default selector is defined!
         spec = self.config.get_selector("default")
     else:
         spec = parse_difference(self.args.models, self.args.exclude) 
     return spec 
```
This works for a selector named `default`! But let's actually do this by defining a new selector property, `default: true|false`, and adding that as an argument to `get_selector`. We'll also want to handle the (very common) case in which a default selector is not defined, by having some way to check for it first.

**Comment by TeddyCr** [comment]
That makes a lot of sense. Thank you for laying this out. Basically from main these are called as follow

`run` -> `_runtime_initialize` -> `selector.get_graph_queue` [`get_node_selector` returns `ResourceTypeSelector`, param passed to `get_graph_queue` `get_selection_spec` returns `selector if args.selector_name`) which will overwrite the graph based on the selector specs `new_graph = self.full_graph.get_subset_graph(selected_nodes)`.

Basically, if I well understand, what we want is if neither `self.args.models` or `self.args.exclude` and there is selector with a `default: true` then overwrite the graph with what is defined in the default selector.

**Comment by jtcohen6** [comment]
That sounds right to me!

**Comment by TeddyCr** [comment]
@jtcohen6 in `project.py` in the `Project` class line 498  is `manifest_selectors` a dict representation of the `selector.yml` file and `selectors` a representation of the nodes to run for each selector name?

I am asking as I would like to implement this feature as shown below -- which requires me to check the `selector.yml` file.

```
IF `selector.yml` is defined THEN
    CHECK if `default` key is define and unique
    IF `default` + unique THEN
        SET spec to the selector name with the default value set
    ELSE not unique
       RAISE error
    ELSE not defined
        RUN regular behavior
    ENDIF
ELSE
    RUN regular behavior
ENDIF
```

**Comment by jtcohen6** [comment]
That makes sense to me. It feels right to raise the error about multiple `default` selectors when the selectors are parsed and validated. Ultimately that's called in `project.py`, but the classes and methods are defined in [`contracts/selectionn.py`](https://github.com/dbt-labs/dbt/blob/e7b8488be8cdc26c6c6da575e5a85213b9f5f267/core/dbt/contracts/selection.py#L8) and [`config/selectors.py`](https://github.com/dbt-labs/dbt/blob/e7b8488be8cdc26c6c6da575e5a85213b9f5f267/core/dbt/config/selectors.py#L32).

Then within the task (codepaths linked above), when deciding whether to use the default selector spec vs. `regular behavior`, the contents of `selectors.yml` are available as `self.config.get_selector`.

**Comment by TeddyCr** [comment]
@jtcohen6  Very useful, I started the implementation. I have updated the `SelectorDefinition` dataclass and implemented the validation in `selector.py`. When I test my implementation in `test_selector_errors.py` to make sure it raises the right error, I am getting a `Could not read selector file data: Additional properties are not allowed ('default' was unexpected)` error which is raised from the hologram validation [[here]](https://github.com/dbt-labs/hologram/blob/0df5196159fad8f23bdcb18119b4337b2d7fa064/hologram/__init__.py#L983). 

My assumption was that hologram was using the dataclass to define the schema to validate against. Is that not accurate?

```python
# selection.py
@dataclass
class SelectorDefinition(dbtClassMixin):
    name: str
    definition: Union[str, Dict[str, Any]]
    description: str = ''
    default: bool = False

@dataclass
class SelectorFile(dbtClassMixin):
    selectors: List[SelectorDefinition]
    version: int = 2
```

**Comment by jtcohen6** [comment]
@TeddyCr Everything you've said above sounds right to me. I don't know why hologram would be raising that error, if you've explicitly defined `default` as a property of the dataclass. You're working from your local clone of dbt (`pip install -r editable-requirements.txt`), right?

**Comment by TeddyCr** [comment]
Yes exactly -- here is my [fork](https://github.com/TeddyCr/dbt/tree/feature/3448-user-defined-default-selection-criteria) and here is what I set up so far [[here]](https://github.com/TeddyCr/dbt/commit/9793c804808376409c8e5d37fe8d263d8700e8ec).

I have my venv where I ran `pip install -r editable-requirements.txt` and then test my unit test like so `pytest -v test/unit/test_selector_errors.py::SelectorUnitTest::test_multiple_default_true`

**Comment by jtcohen6** [comment]
Nice work so far! I was able to check out your fork, and it worked mostly fine. I wasn't able to reproduce the error you're seeing above, but `test_multiple_default_true.py` was failing for a different reason.

I think you want the checks [here](https://github.com/TeddyCr/dbt/blob/9793c804808376409c8e5d37fe8d263d8700e8ec/core/dbt/config/selectors.py#L128) and [here](https://github.com/TeddyCr/dbt/blob/9793c804808376409c8e5d37fe8d263d8700e8ec/core/dbt/config/selectors.py#L132) to be against python `True`, rather than the string `'true'`.

When I made that change and re-ran `test_multiple_default_true`, it passed just fine.

**Comment by TeddyCr** [comment]
Perfect, thanks for the feedback. I’ll continue the implementation then and assume it works (once the small fix is implemented) based on your feedback.

**Comment by TeddyCr** [comment]
Hey @jtcohen6 apologies for the umpteenth question regarding that issue. I have not been providing great value in tackling this issue.

I have an implementation though, I am somewhat stuck in my debugging process.

In `compile.py`, `freshness.py` and `list.py` I have implemented something like below.
https://github.com/TeddyCr/dbt/blob/c0c8df0468b6ae20756518d5f1a42e2832b9fbb6/core/dbt/task/compile.py#L40-L50

Basically, this calls a function `config/project.py` that iterates through the `self.manifest_selectors` which I am assuming is a dictionary of selectors.
https://github.com/TeddyCr/dbt/blob/c0c8df0468b6ae20756518d5f1a42e2832b9fbb6/core/dbt/config/project.py#L656-L666

After implementing this, a bunch of integration test are failing -- which seem to be the results of my implementation. From this, and after trying to figure out what was happening I have 2 questions:
- Is that correct that `self.manifest_selectors` is a dictionary of selectors
- How would I go about doing unit testing against any functions defined in the `Project` class. I am wondering if you folks have an easy way to instantiate a "mock" instance of this class with mock values? I am asking as I would idealy want to test that my function returns what it is supposed to

**Comment by jtcohen6** [comment]
@TeddyCr Nice progress so far! Sorry I was out last week, so just responding now. I have a few quick comments:

### where to get selector info?

I think you want to use `self.selectors` rather than `self.manifest_selectors` (which has already been serialized + expanded, to print a user-friendly version in `manifest.json`). In order to check `self.selectors` for the default selector, you'll need to include `default: True|False` in [`SelectorConfig`](https://github.com/dbt-labs/dbt/blob/459178811bf322c63edddc192a11a55fba8fd101/core/dbt/config/selectors.py#L32), and also the output of the [`parse_from_selectors_definition`](https://github.com/dbt-labs/dbt/blob/459178811bf322c63edddc192a11a55fba8fd101/core/dbt/graph/cli.py#L272-L280) method, since they currently just include the name and definition of each selector.

I think there are a few different ways to go about this. The one I can think of is to change `SelectorConfig` from being just `Dict[str, SelectionSpec]` to instead being `Dict[str, Dict[str, str], Dict[str, SelectionSpec]]`.

For the example selectors in the original comment, instead of:
```python
{'prod': SelectionCriteria(...), 'dev': <dbt.graph.selector_spec.SelectionDifference object at 0x105eb6400>}
```
They would be represented as:
```python
{'prod': {'default': False, 'definition': SelectionCriteria(...)}, 'dev': {'default': True, 'definition': <dbt.graph.selector_spec.SelectionDifference object at 0x105eb6400>}}
```

Then [`get_selector`](https://github.com/dbt-labs/dbt/blob/459178811bf322c63edddc192a11a55fba8fd101/core/dbt/config/project.py#L648-L654) would return `self.selectors[name].definition`, instead of just `self.selectors[name]`.

Alternatively, you could grab the name of the default selector as a pointer earlier on, and store that somewhere. I think you'll need to alter the `selectors` config object one way or another.

### order of operations

Once you can reliably access the default selector definition from the project config, I think you want to slightly adjust the order of the logic you've got above, so that you _do not use_ the default selector if the user has passed `--models` or `--exclude`. How about something like:

```python
    def get_selection_spec(self) -> SelectionSpec:
        # get this first, so we can check to see if it exists in the elif condition below
        default_selector = self.config.get_default_selector()
        logger.info(default_selector)  # TODO remove after debugging
        if self.args.selector_name:
            spec = self.config.get_selector(self.args.selector_name)
        # do not use default selector if the user has passed --models or --exclude,
        # or if it is not defined
        elif not (self.args.models or self.args.exclude) and default_selector:
            spec = default_selector
        else:
            spec = parse_difference(self.args.models, self.args.exclude)
        return spec
```

### testing

The relevant existing unit tests are in [`test_graph_selector_parsing`](https://github.com/dbt-labs/dbt/blob/develop/test/unit/test_graph_selector_parsing.py). If you make changes to the structure of `SelectorConfig`/`selectors` as I mentioned above, it's likely you'll need to adjust this and a few other tests that mock what selectors look like.

I also think we'll want to add an integration test, to make sure this works end-to-end!

Let me know if you find the comments above helpful, and if you're able to give it another go :)

**Comment by TeddyCr** [comment]
@jtcohen6 ah yes, that makes a lot of sense. That was the info I was missing. Sounds good for the integration tests.
